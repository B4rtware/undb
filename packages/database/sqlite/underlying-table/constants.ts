export const INTERNAL_COLUMN_DELETED_AT_NAME = 'deleted_at'

export const M2M_CHILD_ID_FIELD = 'child_id'
export const M2M_PARENG_ID_FIELD = 'parent_id'
