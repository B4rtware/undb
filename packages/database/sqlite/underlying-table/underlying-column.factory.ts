import type { Field } from '@egodb/core'
import type { IUnderlyingColumn } from '../types/underlying-column'
import {
  UnderlyingBoolColumn,
  UnderlyingDateColumn,
  UnderlyingDateRangeFromColumn,
  UnderlyingDateRangeToFromColumn,
  UnderlyingNumberColumn,
  UnderlyingReferenceFromColumn,
  UnderlyingSelectFromColumn,
  UnderlyingStringColumn,
} from './underlying-column'

export class UnderlyingColumnFactory {
  static create(field: Field): IUnderlyingColumn | IUnderlyingColumn[] {
    switch (field.type) {
      case 'string':
        return new UnderlyingStringColumn(field)
      case 'number':
        return new UnderlyingNumberColumn(field)
      case 'bool':
        return new UnderlyingBoolColumn(field)
      case 'date':
        return new UnderlyingDateColumn(field)
      case 'date-range':
        return [new UnderlyingDateRangeFromColumn(field), new UnderlyingDateRangeToFromColumn(field)]
      case 'select':
        return new UnderlyingSelectFromColumn(field)
      case 'reference':
        return new UnderlyingReferenceFromColumn(field)
    }
  }

  static createMany(fields: Field[]): IUnderlyingColumn[] {
    return fields.flatMap((field) => this.create(field))
  }
}
