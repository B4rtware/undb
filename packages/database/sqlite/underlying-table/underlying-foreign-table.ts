import type { Field, ParentField, ReferenceField } from '@egodb/core'
import { INTERNAL_COLUMN_ID_NAME, TreeField } from '@egodb/core'
import type { Knex } from '@mikro-orm/better-sqlite'
import type { IUderlyingForeignTableName, IUnderlyingForeignTable } from '../interfaces/underlying-foreign-table'

abstract class BaseUnderlyingForeignTable<F extends Field> implements IUnderlyingForeignTable {
  constructor(protected readonly foreignTableName: string, protected readonly field: F) {}

  abstract get name(): IUderlyingForeignTableName
  abstract getCreateTableSqls(knex: Knex): string[]
}

export class AdjacencyListTable extends BaseUnderlyingForeignTable<ReferenceField> {
  static CHILD_ID = 'child_id'
  static PARENT_ID = 'parent_id'

  get name(): IUderlyingForeignTableName {
    return `${this.field.id.value}_${this.foreignTableName}_adjacency_list`
  }

  getCreateTableSqls(knex: Knex): string[] {
    return [
      knex.schema
        .createTable(this.name, (tb) => {
          tb.string(AdjacencyListTable.CHILD_ID)
            .notNullable()
            .references(INTERNAL_COLUMN_ID_NAME)
            .inTable(this.foreignTableName)

          tb.string(AdjacencyListTable.PARENT_ID)
            .notNullable()
            .references(INTERNAL_COLUMN_ID_NAME)
            .inTable(this.foreignTableName)

          tb.primary([AdjacencyListTable.CHILD_ID, AdjacencyListTable.PARENT_ID])
        })
        .toQuery(),
    ]
  }
}

export class ClosureTable extends BaseUnderlyingForeignTable<TreeField | ParentField> {
  static CHILD_ID = 'child_id'
  static PARENT_ID = 'parent_id'
  static DEPTH = 'depth'

  get name(): IUderlyingForeignTableName {
    const fieldId = this.field instanceof TreeField ? this.field.id.value : this.field.treeFieldId.value
    return `${fieldId}_${this.foreignTableName}_closure_table`
  }

  getCreateTableSqls(knex: Knex): string[] {
    return [
      knex.schema
        .createTableIfNotExists(this.name, (tb) => {
          tb.string(ClosureTable.CHILD_ID)
            .notNullable()
            .references(INTERNAL_COLUMN_ID_NAME)
            .inTable(this.foreignTableName)
            .onDelete('CASCADE')

          tb.string(ClosureTable.PARENT_ID)
            .notNullable()
            .references(INTERNAL_COLUMN_ID_NAME)
            .inTable(this.foreignTableName)
            .onDelete('CASCADE')

          tb.integer(ClosureTable.DEPTH).notNullable().defaultTo(0)

          tb.primary([ClosureTable.CHILD_ID, ClosureTable.PARENT_ID])
        })
        .toQuery(),
      knex
        .raw(
          `
       create index if not exists \`${this.name}_${ClosureTable.DEPTH}_index\` on \`${this.name}\` (\`${ClosureTable.DEPTH}\`)
       `,
        )
        .toQuery(),
    ]
  }

  connect(knex: Knex, parentRecordId: string, childrenRecordIds: string[] = []): string[] {
    const queries: string[] = []

    const query = knex
      .queryBuilder()
      .table(this.name)
      .delete()
      .where(AdjacencyListTable.PARENT_ID, parentRecordId)
      .toQuery()

    queries.push(query)

    const rootInsert = knex(this.name)
      .insert({
        [ClosureTable.CHILD_ID]: parentRecordId,
        [ClosureTable.DEPTH]: 0,
        [ClosureTable.PARENT_ID]: parentRecordId,
      })
      .toQuery()

    queries.push(rootInsert)

    if (childrenRecordIds?.length) {
      for (const recordId of childrenRecordIds) {
        const query = knex
          .raw(
            `
            insert into
             ${this.name}
              (${ClosureTable.PARENT_ID},
                ${ClosureTable.CHILD_ID},
                ${ClosureTable.DEPTH})

            select
              p.${ClosureTable.PARENT_ID},
              c.${ClosureTable.CHILD_ID},
              p.${ClosureTable.DEPTH}+c.${ClosureTable.DEPTH}+1
            from ${this.name} as p, ${this.name} as c
            where
            p.${ClosureTable.CHILD_ID}='${parentRecordId}'
            and
            c.${ClosureTable.PARENT_ID}='${recordId}'
 `,
          )
          .toQuery()

        queries.push(query)
      }
    }

    return queries
  }

  moveParent(knex: Knex, recordId: string, parentId: string | null): string[] {
    const queries: string[] = []

    queries.push(
      knex
        .queryBuilder()
        .delete()
        .from(this.name)
        .whereIn(
          ClosureTable.CHILD_ID,
          knex.queryBuilder().select(ClosureTable.CHILD_ID).from(this.name).where(ClosureTable.PARENT_ID, recordId),
        )
        .whereNotIn(
          ClosureTable.PARENT_ID,
          knex.queryBuilder().select(ClosureTable.CHILD_ID).from(this.name).where(ClosureTable.PARENT_ID, recordId),
        )
        .toQuery(),
    )

    if (parentId) {
      queries.push(
        `
        INSERT INTO ${this.name} (${ClosureTable.PARENT_ID}, ${ClosureTable.CHILD_ID}, ${ClosureTable.DEPTH})

        SELECT supertree.${ClosureTable.PARENT_ID}, subtree.${ClosureTable.CHILD_ID},
        supertree.${ClosureTable.DEPTH}+subtree.${ClosureTable.DEPTH}+1
        FROM ${this.name} AS supertree JOIN ${this.name} AS subtree
        WHERE subtree.${ClosureTable.PARENT_ID} = '${recordId}'
        AND supertree.${ClosureTable.CHILD_ID} = '${parentId}';
        `,
      )
    }

    return queries
  }
}
