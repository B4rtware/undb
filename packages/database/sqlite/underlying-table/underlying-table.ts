import type { ReferenceField } from '@egodb/core'
import type { IUnderlyingTable } from '../interfaces/underlying-table'

export class UnderlyingM2MTable implements IUnderlyingTable {
  constructor(private readonly tableName: string, private readonly field: ReferenceField) {}

  get name(): string {
    return `${this.field.id.value}_${this.tableName}`
  }
}
