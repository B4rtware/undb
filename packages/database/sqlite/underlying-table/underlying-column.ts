import type {
  BoolField,
  ColorField,
  DateField,
  DateRangeField,
  EmailField,
  Field,
  NumberField,
  ParentField,
  ReferenceField,
  SelectField,
  StringField,
  TreeField,
} from '@egodb/core'
import {
  INTERNAL_COLUMN_CREATED_AT_NAME,
  INTERNAL_COLUMN_ID_NAME,
  INTERNAL_COLUMN_UPDATED_AT_NAME,
  INTERNAL_INCREAMENT_ID_NAME as INTERNAL_AUTO_INCREAMENT_ID_NAME,
} from '@egodb/core'
import type { Knex } from '@mikro-orm/better-sqlite'
import type { IUnderlyingColumn } from '../interfaces/underlying-column'
import { INTERNAL_COLUMN_DELETED_AT_NAME } from './constants'

export abstract class UnderlyingColumn implements IUnderlyingColumn {
  abstract get name(): string
  abstract build(tb: Knex.TableBuilder, knex: Knex): void
}

export class UnderlyingIdColumn extends UnderlyingColumn {
  get name(): string {
    return INTERNAL_COLUMN_ID_NAME
  }

  build(tb: Knex.TableBuilder): void {
    tb.string(this.name).notNullable()
  }
}

export class UnderlyingAutoIncreamentColumn extends UnderlyingColumn {
  get name(): string {
    return INTERNAL_AUTO_INCREAMENT_ID_NAME
  }

  build(tb: Knex.TableBuilder): void {
    tb.increments(this.name).notNullable()
  }
}

export class UnderlyingCreatedAtColumn extends UnderlyingColumn {
  get name(): string {
    return INTERNAL_COLUMN_CREATED_AT_NAME
  }

  build(tb: Knex.TableBuilder, knex: Knex): void {
    tb.datetime(this.name).notNullable().defaultTo(knex.fn.now())
  }
}

export class UnderlyingUpdatedAtColumn extends UnderlyingColumn {
  get name(): string {
    return INTERNAL_COLUMN_UPDATED_AT_NAME
  }

  build(tb: Knex.TableBuilder, knex: Knex): void {
    tb.datetime(this.name).notNullable().defaultTo(knex.fn.now())
  }
}

export class UnderlyingDeletedAtColumn extends UnderlyingColumn {
  get name(): string {
    return INTERNAL_COLUMN_DELETED_AT_NAME
  }

  build(tb: Knex.TableBuilder): void {
    tb.datetime(this.name).nullable()
  }
}

abstract class UnderlyingFieldColumn<F extends Field> implements IUnderlyingColumn {
  constructor(protected readonly field: F) {}
  get name(): string {
    return this.field.id.value
  }
  abstract build(tb: Knex.TableBuilder, knex: Knex): void
}

export class UnderlyingStringColumn extends UnderlyingFieldColumn<StringField> {
  build(tb: Knex.TableBuilder): void {
    tb.string(this.name)
  }
}

export class UnderlyingEmailColumn extends UnderlyingFieldColumn<EmailField> {
  build(tb: Knex.TableBuilder): void {
    tb.string(this.name)
  }
}

export class UnderlyingColorColumn extends UnderlyingFieldColumn<ColorField> {
  build(tb: Knex.TableBuilder): void {
    tb.string(this.name, 10)
  }
}

export class UnderlyingNumberColumn extends UnderlyingFieldColumn<NumberField> {
  build(tb: Knex.TableBuilder): void {
    tb.float(this.name)
  }
}

export class UnderlyingBoolColumn extends UnderlyingFieldColumn<BoolField> {
  build(tb: Knex.TableBuilder): void {
    tb.boolean(this.name)
  }
}

export class UnderlyingDateColumn extends UnderlyingFieldColumn<DateField> {
  build(tb: Knex.TableBuilder): void {
    tb.dateTime(this.name)
  }
}

export class UnderlyingDateRangeFromColumn extends UnderlyingFieldColumn<DateRangeField> {
  get name(): string {
    return super.name + '_from'
  }

  build(tb: Knex.TableBuilder): void {
    tb.dateTime(this.name)
  }
}

export class UnderlyingDateRangeToFromColumn extends UnderlyingFieldColumn<DateRangeField> {
  get name(): string {
    return super.name + '_to'
  }

  build(tb: Knex.TableBuilder): void {
    tb.dateTime(this.name)
  }
}

export class UnderlyingSelectFromColumn extends UnderlyingFieldColumn<SelectField> {
  build(tb: Knex.TableBuilder): void {
    tb.string(this.name)
  }
}

export class UnderlyingReferenceColumn extends UnderlyingFieldColumn<ReferenceField> {
  build(tb: Knex.TableBuilder): void {
    tb.json(this.name)
  }
}

export class UnderlyingTreeColumn extends UnderlyingFieldColumn<TreeField> {
  build(tb: Knex.TableBuilder): void {
    tb.json(this.name)
  }
}

export class UnderlyingParentColumn extends UnderlyingFieldColumn<ParentField> {
  build(tb: Knex.TableBuilder): void {
    tb.string(this.name).nullable()
  }
}
