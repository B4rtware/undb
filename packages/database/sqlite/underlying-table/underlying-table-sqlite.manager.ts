import type { ITableSpec, IUnderlyingTableManager, Table } from '@egodb/core'
import { BaseEntityManager } from '../repository/base-entity-manager'
import { UnderlyingTableSqliteManagerVisitor } from './underlying-table-sqlite.manager-visitor'
import { UnderlyingTableBuilder } from './underlying-table.builder'

export class UnderlyingTableSqliteManager extends BaseEntityManager implements IUnderlyingTableManager {
  async create(table: Table): Promise<void> {
    const queries = new UnderlyingTableBuilder(this.em).createTable(table).build()

    this.addQueries(...queries)

    await this.commit()
  }

  async update(tableId: string, spec: ITableSpec): Promise<void> {
    const visitor = new UnderlyingTableSqliteManagerVisitor(tableId, this.em)
    spec.accept(visitor)

    await visitor.commit()
  }

  async delete(tableId: string): Promise<void> {
    // TODO: drop table later
    // const knex = this.em.getKnex()
    // const sb = knex.schema.dropTable(tableId)
    // await this.em.execute(sb.toQuery())
  }
}
