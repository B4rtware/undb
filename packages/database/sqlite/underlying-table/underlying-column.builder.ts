import type { Field } from '@egodb/core'
import { INTERNAL_COLUMN_ID_NAME, INTERNAL_COLUMN_UPDATED_AT_NAME } from '@egodb/core'
import type { Knex } from '@mikro-orm/better-sqlite'
import type { IUnderlyingColumnBuilder } from '../interfaces/underlying-table.builder'
import {
  UnderlyingAutoIncreamentColumn,
  UnderlyingCreatedAtColumn,
  UnderlyingDeletedAtColumn,
  UnderlyingIdColumn,
  UnderlyingUpdatedAtColumn,
} from './underlying-column'
import { UnderlyingColumnFactory } from './underlying-column.factory'

export class UnderlyingColumnBuilder implements IUnderlyingColumnBuilder {
  constructor(private readonly knex: Knex, private readonly tb: Knex.TableBuilder) {}
  private queries: string[] = []

  private addQueries(...queries: string[]) {
    for (const query of queries) {
      this.queries.push(query)
    }
  }

  public build() {
    return this.queries
  }

  createAutoIncrement(): this {
    new UnderlyingAutoIncreamentColumn().build(this.tb)

    return this
  }

  createId(tableName: string): this {
    const column = new UnderlyingIdColumn()
    column.build(this.tb)

    const index = this.knex
      .raw(
        `
      create index \`${tableName}_${column.name}_index\` on \`${tableName}\` (\`${column.name}\`)
    `,
      )
      .toQuery()

    const unique = this.knex
      .raw(
        `
      create unique index \`${tableName}_${column.name}_unique\` on \`${tableName}\` (\`${column.name}\`)
      `,
      )
      .toQuery()

    this.addQueries(index, unique)

    return this
  }

  createCreatedAt(): this {
    new UnderlyingCreatedAtColumn().build(this.tb, this.knex)
    return this
  }

  createUpdatedAt(tableName: string): this {
    new UnderlyingUpdatedAtColumn().build(this.tb, this.knex)

    const query = this.knex
      .raw(
        `
		CREATE TRIGGER update_at_update_${tableName} AFTER UPDATE ON \`${tableName}\`
		BEGIN
			update \`${tableName}\` SET ${INTERNAL_COLUMN_UPDATED_AT_NAME} = datetime('now') WHERE ${INTERNAL_COLUMN_ID_NAME} = NEW.${INTERNAL_COLUMN_ID_NAME};
		END;
	 `,
      )
      .toQuery()

    this.addQueries(query)

    return this
  }

  createDeletedAt(): this {
    new UnderlyingDeletedAtColumn().build(this.tb)
    return this
  }

  createUnderlying(fields: Field[]): this {
    const underlyingColumns = UnderlyingColumnFactory.createMany(fields)

    for (const column of underlyingColumns) {
      column.build(this.tb, this.knex)
    }

    return this
  }
}
