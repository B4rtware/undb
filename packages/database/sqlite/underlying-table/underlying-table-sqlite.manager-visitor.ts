/* eslint-disable @typescript-eslint/no-empty-function */
import type { ITableSpecVisitor, WithNewField, WithoutField, WithoutOption, WithTableSchema } from '@egodb/core'
import type { EntityManager, Knex } from '@mikro-orm/better-sqlite'
import { UnderlyingColumnBuilder } from './underlying-column.builder'
import { UnderlyingColumnFactory } from './underlying-column.factory'

export class UnderlyingTableSqliteManagerVisitor implements ITableSpecVisitor {
  private readonly knex: Knex
  private sb?: Knex.SchemaBuilder
  private qb?: Knex.QueryBuilder
  private sqls: string[] = []
  constructor(private readonly tableName: string, private readonly em: EntityManager) {
    const knex = em.getKnex()
    this.knex = knex
  }

  get #sb() {
    return this.sb ?? this.knex.schema
  }

  get #qb() {
    return this.qb ?? this.knex.queryBuilder()
  }

  public get queries(): string[] {
    return [this.sb?.toQuery() ?? '', this.qb?.toQuery() ?? '', ...this.sqls].filter(Boolean)
  }

  async commit() {
    for (const query of this.queries) {
      await this.em.execute(query)
    }
  }

  idEqual(): void {
    throw new Error('Method not implemented.')
  }
  nameEqual(): void {}
  schemaEqual(s: WithTableSchema): void {
    this.sb = this.#sb.alterTable(this.tableName, (tb) => {
      const builder = new UnderlyingColumnBuilder(this.knex, tb)
      builder.createUnderlying(s.schema.fields)
    })
  }
  viewsEqual(): void {}
  viewEqual(): void {}
  filterEqual(): void {}
  newField(s: WithNewField): void {
    this.sb = this.#sb.alterTable(this.tableName, (tb) => {
      const builder = new UnderlyingColumnBuilder(this.knex, tb)
      builder.createUnderlying([s.field])
    })
  }
  fieldsOrder(): void {}
  fieldWidthEqual(): void {}
  fieldVisibility(): void {}
  displayTypeEqual(): void {}
  kanbanFieldEqual(): void {}
  calendarFieldEqual(): void {}
  optionsEqual(): void {}
  newOption(): void {}
  optionEqual(): void {}
  witoutOption(s: WithoutOption): void {
    this.qb = this.#qb.from(this.tableName).where(s.field.id.value, s.optionKey.value).update(s.field.id.value, null)
  }
  withoutField(s: WithoutField): void {
    const fields = UnderlyingColumnFactory.createMany([s.field])
    const sqls = fields.map(
      (f) =>
        `
    alter table \`${this.tableName}\` drop column \`${f.name}\`;
    `,
    )
    this.sqls.push(...sqls)
  }
  fieldOptionsEqual(): void {}
  not(): this {
    return this
  }
}
