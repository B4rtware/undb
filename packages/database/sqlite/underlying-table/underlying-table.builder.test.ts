import {
  createTestTable,
  FieldId,
  FieldName,
  FieldValueConstraints,
  StringField,
  TableSchema,
  WithTableSchema,
} from '@egodb/core'
import { EntityManager } from '@mikro-orm/better-sqlite'
import { UnderlyingTableBuilder } from './underlying-table.builder'

describe('UnderlyingTableBuilder', () => {
  let em: EntityManager

  beforeAll(() => {
    // @ts-expect-error
    em = global.em
  })

  test('should build table', () => {
    const queries = new UnderlyingTableBuilder(em)
      .createTable(
        createTestTable(
          new WithTableSchema(
            new TableSchema([
              new StringField({
                id: FieldId.fromString('fldid'),
                name: FieldName.create('name'),
                valueConstrains: FieldValueConstraints.create({}),
              }),
            ]),
          ),
        ),
      )
      .build()

    expect(queries).toMatchSnapshot()
  })
})
