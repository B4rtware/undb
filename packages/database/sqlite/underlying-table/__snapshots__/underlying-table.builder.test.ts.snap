// Vitest Snapshot v1

exports[`UnderlyingTableBuilder > should build table 1`] = `
[
  "create table \`tableId\` (\`auto_increment\` integer not null primary key autoincrement, \`id\` varchar(255) not null, \`created_at\` datetime not null default CURRENT_TIMESTAMP, \`updated_at\` datetime not null default CURRENT_TIMESTAMP, \`deleted_at\` datetime null, \`fldid\` varchar(255))",
  "
		CREATE TRIGGER update_at_update_tableId AFTER UPDATE ON \`tableId\`
		BEGIN
			update \`tableId\` SET updated_at = datetime('now') WHERE id = NEW.id;
		END;
	 ",
  "
      create unique index \`tableId_id_unique\` on \`tableId\` (\`id\`)
      ",
]
`;
